// @app/models/Country.js
import BaseModel from './BaseModel';
// TODO: Change all to api or modelApi, easier to copy paste the file
import * as api from '@app/api/country';
import moment from 'moment';

export default class CountryModel extends BaseModel {

    constructor({ model = {}, token }) {
        super({ model, token, api });
        this.attributes = {
            id: null,
            name: null,
            iso2: null,
            iso3: null,
            ...model
        };
        // safe attributes to send for update/create API
        this.safeAttributes = ['name', 'iso2', 'iso3'];
        // attributes label for display, for active inputs
        // TODO: integrate with validate.js lib for error message
        this.labels = {
            id: 'ID',
            name: 'Name',
            iso2: 'ISO2',
            iso3: 'ISO3',
        };
        // check validate.js
        this.rules = {
            name: {
                type: "string",
                presence: { allowEmpty: false },
            },            
            iso2: {
                type: "string",
                presence: { allowEmpty: false },
                length: { is: 2 }
            },
            iso3: {
                type: "string",
                presence: { allowEmpty: false },
                length: { is: 3 }
            },            
        };
    }

}
