// @app/models/Holiday.js
import BaseModel from './BaseModel';
import * as api from '@app/api/holiday';
import { 
    STATUS_ACTIVE, 
    STATUS_DELETED
} from '@app/constants';

export default class HolidayModel extends BaseModel {

    constructor({ model = {}, token }) {
        super({ model, token, api });
        this.attributes = {
            id: null,
            title: null,
            start: null,
            end: null,
            start_ts: null,
            end_ts: null,
            status: null,
            country_id: null,
            ...model
        };
        // safe attributes to send for update/create API
        this.safeAttributes = ['title', 'start', 'end', 'status', 'country_id'];
        // attributes label for display, for active inputs
        // TODO: integrate with validate.js lib for error message
        this.labels = {
            id: 'ID',
            title: 'Title',
            start: 'Start Date',
            end: 'End Date',
            start_ts: 'Start Timestamp',
            end_ts: 'End Timestamp',
            status: 'Status',
            country_id: 'Country',

            created_at: 'Created At',
            created_by: 'Created By',
            updated_at: 'Updated At',
            updated_by: 'Updated By',
        };
        // check validate.js
        this.rules = {
            country_id: {
                presence: { allowEmpty: false },
                numericality: {
                    onlyInteger: true,
                    greaterThan: 0,
                },
            },
            title: {
                type: "string",
                presence: { allowEmpty: false },
            },
            start: {
                presence: { allowEmpty: false },
                datetime: { dateOnly: true },
            },
            end: {
                presence: { allowEmpty: false },
                datetime: { dateOnly: true },
            },
            status: {
                type: "string",
                presence: { allowEmpty: false },
                inclusion: [STATUS_ACTIVE, STATUS_DELETED],
            },
        };
        // all others options
        this.statusOptions = [
            { value: STATUS_ACTIVE, label: STATUS_ACTIVE },
            { value: STATUS_DELETED, label: STATUS_DELETED },
        ];
    }

    getCountry() {
        return this.attributes.country;
    }

}
