// @app/api/index.js
import axios from 'axios';

const BASE_URL = QA_API_URL;

async function send({ method, path, data, token, header = {} }) {
    console.log(`API: ${method} ${path}`);
    try {
        const response = await axios({
            method,
            baseURL: BASE_URL,
            url: path,
            data,
            headers: {
                ...header,
                ...(token ? { Authorization: `Bearer ${token}` } : {})
            }
        });
        return new Promise(resolve => resolve(response.data));
    } catch (error) {
        throw error;
    }
}

export function post(path, data, token = null, header = {}) {
    return send({ method: 'post', path, data, token, header });
}

export function put(path, data, token = null, header = {}) {
    return send({ method: 'put', path, data, token, header });
}

export function del(path, token = null, header = {}) {
    return send({ method: 'delete', path, data: {}, token, header });
}

export function get(path, data = {}, token = null, header = {}) {
    const params = Object.keys(data).map(key => `${key}=${data[key]}`).join('&');
    return send({
        method: 'get',
        path: `${path}${params ? '?' + params : ''}`,
        token,
        header
    });
}
