// @app/api/user.js
import * as api from '@app/api'

const urlPrefix = 'user';

export function checkAccess(token) {
    return api.get(`${urlPrefix}/check-access`, [], token);
}

export function register(email, password, job_role, display_name) {
    return api.post(`${urlPrefix}/create`, { email, password, job_role, display_name })
}

export function login(email, password) {
    return api.post(`${urlPrefix}/auth`, { email, password });
}

export function logout(token) {
    return api.post(`${urlPrefix}/logout`, [], token);
}

export function view(id, data, token) {
    return api.get(`${urlPrefix}/${id}`, data, token);
}

export function update(id, data, token) {
    return api.put(`${urlPrefix}/${id}`, data, token);
}

export function remove(id, token) {
    return api.del(`${urlPrefix}/${id}`, token);
}

export function list(data, token) {
    return api.get(`${urlPrefix}`, data, token);
}

export function getActiveUserList(token) {
    return api.get(`${urlPrefix}`, {
        fields: 'id,display_name,job_role,status,created_at',
        sort: 'display_name',
        status: 'Active',
        'per-page': 500
    }, token);
}

export function changePassword(id, oldPassword, newPassword, token) {
    return api.put(
        `${urlPrefix}/change-password/${id}`, 
        {
            'password_new': newPassword, 
            'password_current': oldPassword
        },
        token
    );
}

export function forgetPassword(email) {
    return api.post(`${urlPrefix}/forget-password`, { email });
}

export function resetPassword(password, passwordResetToken) {
    return api.post(`${urlPrefix}/reset-password`, { password, passwordResetToken });
}

export function assignRole(id, role, token) {
    return api.put(`${urlPrefix}/assign-role/${id}`, { role }, token);
}

export function revokeRole(id, role, token) {
    return api.put(`${urlPrefix}/revoke-role/${id}`, { role }, token);
}
