// @app/constants.js
export const STATUS_ACTIVE = 'Active'
export const STATUS_INACTIVE = 'Inactive'
export const STATUS_DEPLOYED = 'Deployed'
export const STATUS_DELETED = 'Deleted'

export const TEST_STATUS_PASSED = 'Passed'
export const TEST_STATUS_FAILED = 'Failed'
export const TEST_STATUS_PENDING = 'Pending'
export const TEST_STATUS_UNABLE_TO_TEST = 'Unable To Test'

export const JOB_ROLE_USER = 'User';
export const JOB_ROLE_DEVELOPER = 'Developer';
export const JOB_ROLE_QA = 'QA';
export const JOB_ROLE_PM = 'PM';
export const JOB_ROLE_MANAGEMENT = 'Management';

export const ROLE_USER = 'User';
export const ROLE_ADMIN = 'Admin';
export const ROLE_TIMELINE = 'Timeline';

export const FIX_STATUS_PENDING_INVESTIGATION = 'Pending Investigation';
export const FIX_STATUS_UNABLE_TO_REPLICATE = 'Unable to Replicate';
export const FIX_STATUS_NEED_MORE_INFORMATION = 'Need More Information';
export const FIX_STATUS_WORK_IN_PROGRESS = 'Work In Progress';
export const FIX_STATUS_TO_DO = 'To Do';
export const FIX_STATUS_FIXED = 'Fixed';
export const FIX_STATUS_KIV = 'KIV';

export const PRIORITY_LOW = 'Low';
export const PRIORITY_MEDIUM = 'Medium';
export const PRIORITY_HIGH = 'High';
export const PRIORITY_BLOCKER = 'Blocker';

export const TYPE_BUG = 'Bug';
export const TYPE_FEATURE = 'Feature';
